library/psa_crypto.c

```diff
@@ -4069,16 +4069,42 @@ static psa_status_t psa_cipher_setup( psa_cipher_operation_t *operation,
                               PSA_KEY_USAGE_ENCRYPT :
                               PSA_KEY_USAGE_DECRYPT );
 
+    status = psa_get_key_from_slot( handle, &slot, usage, alg );
+    if( status != PSA_SUCCESS )
+        goto exit;
+
     /* A context must be freshly initialized before it can be set up. */
     if( operation->alg != 0 )
-    {
         return( PSA_ERROR_BAD_STATE );
-    }
 
     status = psa_cipher_init( operation, alg );
     if( status != PSA_SUCCESS )
         return( status );
 
+    /* Try doing this through a driver before using software fallback */
+    if( cipher_operation == MBEDTLS_ENCRYPT )
+        status = psa_driver_wrapper_cipher_encrypt_setup( operation,
+                                                          slot,
+                                                          alg );
+    else
+        status = psa_driver_wrapper_cipher_decrypt_setup( operation,
+                                                          slot,
+                                                          alg );
+
+    if( status == PSA_SUCCESS )
+    {
+        operation->accelerator_set = 1;
+        operation->key_set = 1;
+    }
+
+    if( status != PSA_ERROR_NOT_SUPPORTED ||
+        psa_key_lifetime_is_external( slot->attr.lifetime ) )
+        goto exit;
```

Comment:
If the status from the driver setup call is `NOT_SUPPORTED`, and the key is not external (i.e. it is transparent), then the code will fall through to do `mbedtls_cipher_init` and the rest of the initialisation of the software fallback, all the while `accelerator_set` is rightfully 0.

I don't see the issue? Am I missing something, or are you?

Reply:
> When the setup wrapper returns a failure code, it hasn't touched the `operation->ctx` union, and it will therefore still be nulled.

Yes, and so `mbedtls_cipher_free` is given an input of all-bits-zero, which is not guaranteed to be valid.

Reply:
Fixed and documented now.